autopilot/contract_pruning.go

```diff
@@ -153,102 +112,92 @@ func (ap *Autopilot) performContractPruning(wp *workerPool) {
-			}
-			cancel()
-
-			// handle metrics
-			metrics = append(metrics, result.toMetric())
+	// dismiss alerts for contracts that are no longer prunable
+	ap.dismissPruneAlerts(prunable)
+
+	// loop prunable contracts
+	var total uint64
+	for _, contract := range prunable {
+		// check if we're stopped
+		if ap.isStopped() {
+			break
 		}
-	})
 
-	// record metrics
-	ctx, cancel := context.WithTimeout(ap.shutdownCtx, time.Minute)
-	if err := ap.bus.RecordContractPruneMetric(ctx, metrics...); err != nil {
-		ap.logger.Error(err)
+		// fetch host
+		h, _, err := ap.fetchHostContract(contract.ID)
+		if utils.IsErr(err, api.ErrContractNotFound) {
+			continue // contract got archived
+		} else if err != nil {
+			ap.logger.Errorf("failed to fetch host for contract '%v', err %v", contract.ID, err)
+			continue
+		}
+
+		// prune contract using a random worker
+		wp.withWorker(func(w Worker) {
+			total += ap.pruneContract(w, contract.ID, h.PublicKey, h.Settings.Version)
+		})
 	}
-	cancel()
 
-	// log metrics
-	ap.logger.Info(metrics)
+	// log total pruned
+	ap.logger.Info(fmt.Sprintf("pruned %d (%s) from %v contracts", total, humanReadableSize(int(total)), len(prunable)))
 }
 
-func (ap *Autopilot) pruneContract(w Worker, fcid types.FileContractID) pruneResult {
-	// create a sane timeout
-	ctx, cancel := context.WithTimeout(ap.shutdownCtx, 2*timeoutPruneContract)
+func (ap *Autopilot) pruneContract(w Worker, fcid types.FileContractID, hk types.PublicKey, hostVersion string) uint64 {
+	// use a sane timeout
+	ctx, cancel := context.WithTimeout(ap.shutdownCtx, timeoutPruneContract+5*time.Minute)
 	defer cancel()
 
-	// fetch the host
-	host, _, err := ap.hostForContract(ctx, fcid)
-	if err != nil {
-		return pruneResult{fcid: fcid, err: err}
-	}
-
 	// prune the contract
 	start := time.Now()
 	pruned, remaining, err := w.RHPPruneContract(ctx, fcid, timeoutPruneContract)
-	if err != nil && pruned == 0 {
-		return pruneResult{fcid: fcid, hk: host.PublicKey, version: host.Settings.Version, err: err}
-	} else if err != nil && utils.IsErr(err, context.DeadlineExceeded) {
+	duration := time.Since(start)
+
+	// ignore slow pruning until host network is 1.6.0+
+	if utils.IsErr(err, context.DeadlineExceeded) && pruned > 0 {
 		err = nil
 	}
 
-	return pruneResult{
-		ts: start,
-
-		fcid:    fcid,
-		hk:      host.PublicKey,
-		version: host.Settings.Version,
+	// handle metrics
+	if err == nil || pruned > 0 {
+		if err := ap.bus.RecordContractPruneMetric(ctx, api.ContractPruneMetric{
+			Timestamp: api.TimeRFC3339(start),
 
-		pruned:    pruned,
-		remaining: remaining,
-		duration:  time.Since(start),
+			ContractID:  fcid,
+			HostKey:     hk,
+			HostVersion: hostVersion,
 
-		err: err,
+			Pruned:    pruned,
+			Remaining: remaining,
+			Duration:  duration,
+		}); err != nil {
+			ap.logger.Error(err)
+		}
 	}
-}
 
-func humanReadableSize(b int) string {
-	const unit = 1024
-	if b < unit {
-		return fmt.Sprintf("%d B", b)
+	// handle logs
```

Comment:
I'd use `logger.With` here for `contract`, `host`, `version`, `pruned`, `remaining`, `elapsed` and `err` and then just conditionally call `Error` or `Info` to make better use of the structured logging rather than squeezing it into a string.